### ISSUE TITLE
Combined BlocBuilder/Listener

### ISSUE THREAD

**Issue by fl0cke** [felangel/bloc#545]
**Is your feature request related to a problem? Please describe.**
I often find myself in the situation where I need both bloc builders and bloc listeners in the same widget tree, e.g. for showing a loading indicator (builder) and displaying a snackbar on success (listener).

**Describe the solution you'd like**
A BlocBuilderListener Widget that accepts both a builder function and a listener would greatly reduce the clutter in my code and also reduce the overhead of having two stateful widgets that basically do the same. Alternatively, you could add a listener parameter to the existing BlocBuilder widget.

**Comment by felangel** [comment]
Hi @fl0cke 👋 
Thanks for opening an issue!

Can you please elaborate a bit on why you say this would reduce the clutter in your code?
If I'm understanding correctly, you're proposing to go from:

```dart
BlocListener<MyBloc, MyState>(
  listener: (context, state) { ... },
  child: BlocBuilder<MyBloc, MyState>(
    builder: (context, state) { ... }
  ),
),
```

to

```dart
BlocBuilderListener<MyBloc, MyState>(
  listener: (context, state) { ... },
  builder: (context, state) { ... },
),
```

I personally don't think it's cluttered as is and I'd be more concerned about confusing people because it isn't immediately obvious what `BlocBuilderListener` would do.

Thoughts?

**Comment by fl0cke** [comment]
Hey @felangel,
Thanks for the quick reply!

Yes, that's exactly what the proposed BlocBuilderListener would look like. Alternatively, we could add an optional `listener` parameter to the `BlocBuilder` to avoid introducing a new widget. 

I agree that it's not that much clutter, but when i look at my usage of the two BlocX widgets, i often find myself using either just a `BlocBuilder` or a `BlocBuilder` nested in a `BlocListener`, but almost never a `BlocListener` on its own. So this would be good opportunity to reduce the number of widgets that are required to achieve a reactive UI + side effects.

Example: A common use case is submitting some sort of user input and displaying a loading indicator while doing so. On success, I want to both show a SnackBar (using the listener) and hide the loading indicator (using the builder).

**Comment by bigworld12** [comment]
I would suggest the merge to have the same signature as `ListView` and `ListView.builder`

e.g.

```dart
//normal BlocListener, listener is triggered on state change, child doesn't get rebuilt on state change
BlocView<MyBloc,MyState>(
  listener: (context, state) { ... }, 
  child: Container()
)

//works as BLocBuilder and BlocListener combined, listener and builder get triggered on state change
BlocView<MyBloc,MyState>.builder(
  listener: (context, state) { ... }, //listener here is optional
  builder: (context, state) { ... },
)
```
(this also goes well with my proposal here https://github.com/felangel/bloc/issues/560)
e.g.
```dart
BlocView<MyBloc>(
  listener: (context, state) { ... }, 
  child: Container()
)

BlocView<MyBloc>.builder(
  listener: (context, state) { ... }, //listener here is optional
  builder: (context, state) { ... },
)
```

**Comment by RollyPeres** [comment]
I definitely think that the `listener` should be added to the `BlocBuilder` or `BlocView`, whatever you prefer as a name. You would only have to differentiate the build/listen condition, which could be two separate params like `buildWhen` and `listenWhen`. A lot cleaner approach and extremely helpful overall.

**Comment by felangel** [comment]
@RollyPeres but you’d also still have a separate BlocListener in cases where you only want to listen?

**Comment by RollyPeres** [comment]
That or make the `builder` optional. But that would probably require a name change to something more generic not strictly tied to building, maybe `BlocView` or something.

**Comment by felangel** [comment]
BlocView still implies there is a view so it also wouldn’t make sense to have an optional builder imo. I personally don’t want to over complicate and muddy the role of BlocBuilder which is why I’ve been hesitant to introduce this. The benefit of introducing this is not that significant imo. Thoughts?

**Comment by RollyPeres** [comment]
I agree in regards to naming, I just threw the name which @bigworld12 came up with for the sake of example. Well, I also understand there are many of your users already comfortable with the current state of this library. The smartest way to go about this would probably be to keep those two widgets as they are and introduce a new one which combines the two together and let users pick and choose ✌

**Comment by felangel** [comment]
Yeah that’s what I was leaning toward as well but not sure what a good, obvious name would be 🤔

**Comment by RollyPeres** [comment]
`BlocConsumer`  ? RIP Remi's naming conventions 👍

**Comment by felangel** [comment]
That could work haha. I’ll think about it some more and discuss with the team 👍

**Comment by Cgaafary** [comment]
Just a few other names to throw out there for consideration... BlocClient,
BlocReceiver, and BlocRetriever.  Not sure if that would violate other
naming conventions. I think both BlocClient and BlocConsumer could imply a
generalist role of doing something with provided blocs. Makes it easier to
add additional functionality in the future with optional parameters in my
opinion. Also, BlocConsumer and BlocReceiver or BlocRetriever feel
intuitively antagonistic to BlocProvider in a general since.

**Comment by Cgaafary** [comment]
Just a few additional names to consider are BlocReceiver and BlocRetriever. 

I think also linguistically it feels like something that Provides aka a "BlocProvider" should be matched with something that receives or retrieves. Once a Bloc is received or retrieved, anything could be done with it such as building a widget or listening to the bloc to perform another task. This naming convention could potentially make it easier to add additional functionality in the future through the use of optional parameters. 

A consumer linguistically feels more opposite a to a "producer" rather than a "provider." BlocConsumer however may fit already more familiar naming conventions with Dart/Flutter.  

I also considered BlocClient but that seems a little less intuitive to me as something that receives from a BlocProvider. It feels more appropriate that a BlocClient would be opposite to a BlocServer to fit more traditional naming conventions.

**Comment by RollyPeres** [comment]
Since `bloc` has a small flavor of redux, I'll also nominate `BlocConnector`.

**Comment by hawkinsjb1** [comment]
Whether or not this is still being considered, i'm going to spitball a few names for the fun of it

BlocWidget, BlocComponent, BlocBuilder.listener(), BlocReceiver is cooler than BlocConsumer IMO (and pairs with BlocProvider), BlocBase, BlocPost, BlocStation, BlocReactor, ehh? I think just BlocWidget makes sense as far as BLoC pattern goes, logical classes represented by Bloc<Event, State> paired with a BlocWidget that presents the interface

**Comment by felangel** [comment]
Implemented in #753 and released in `flutter_bloc v3.1.0` 🎉
